compiler/rustc_lexer/src/unescape.rs

```diff
@@ -379,30 +381,42 @@ where
     }
 }
 
+/// Skip ASCII whitespace, except for the formfeed character
+/// (see [this issue](https://github.com/rust-lang/rust/issues/136600)).
+/// Warns on unescaped newline and following non-ASCII whitespace.
 fn skip_ascii_whitespace<F>(chars: &mut Chars<'_>, start: usize, callback: &mut F)
 where
     F: FnMut(Range<usize>, EscapeError),
 {
-    let tail = chars.as_str();
-    let first_non_space = tail
-        .bytes()
-        .position(|b| b != b' ' && b != b'\t' && b != b'\n' && b != b'\r')
-        .unwrap_or(tail.len());
-    if tail[1..first_non_space].contains('\n') {
-        // The +1 accounts for the escaping slash.
-        let end = start + first_non_space + 1;
+    let mut spaces = 0;
+    let mut contains_nl = false;
+
+    for byte in chars.as_str().bytes() {
+        match byte {
+            b' ' | b'\t' | b'\r' => {
+                spaces += 1;
+            }
+            b'\n' => {
+                spaces += 1;
+                contains_nl = true;
+            }
+            _ => break,
+        }
+    }
```

Comment:
if performance is the goal, it might be advantageous to make this loop use fewer branches. 

I believe this is equivalent, but uses fewer instructions and branches https://godbolt.org/z/boTz6s7bT

```suggestion
    for &byte in chars.as_str().bytes() {
        let is_space = matches!(byte, b' ' | b'\t' | b'\r' | b'\n');
        let is_newline = byte == b'\n';

        spaces += is_space as usize;
        contains_nl |= is_newline;

        if !is_space {
            break;
        }
    }
```

Reply:
It looks like the match in my version creates a big jumptable. Your version looks very promising.